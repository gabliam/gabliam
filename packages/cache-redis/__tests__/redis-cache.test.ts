import RedisCache from '../src/index';
import * as Redis from 'ioredis';

let cache: RedisCache;
let cache2: RedisCache;
let cache3: RedisCache;

beforeAll(async () => {
  cache = new RedisCache('test');
  cache2 = new RedisCache('test2', {
    duration: 2,
    mode: 'EX',
  });
  cache3 = new RedisCache('test2', {
    timeout: 500,
  });
  await cache.start();
  await cache2.start();
  await cache3.start();
});

afterEach(async () => {
  await cache.clear();
  await cache2.clear();
  await cache3.clear();
});

afterAll(async () => {
  await cache.stop();
  await cache2.stop();
  await cache3.stop();
});

test('cache', () => {
  expect(cache.getName()).toMatchSnapshot();
  expect(cache.getNativeCache()).toBeInstanceOf(RedisCache);
});

test('cache get & put', async () => {
  expect(await cache.get('test')).toMatchSnapshot();
  await cache.put('test', 'test');
  expect(await cache.get('test')).toMatchSnapshot();
  expect(await cache.get('test')).toMatchSnapshot();
  await cache.put('test', undefined);
  expect(await cache.get('test')).toMatchSnapshot();
  await cache.put('test2', null);
  expect(await cache.get('test2')).toMatchSnapshot();
});

test('cache putIfAbsent', async () => {
  await cache.put('test', 'test');
  expect(await cache.putIfAbsent('test', 'testnew')).toMatchSnapshot();
  expect(await cache.putIfAbsent('test2', 'test2')).toMatchSnapshot();
  expect(await cache.putIfAbsent('test3', 'test3')).toMatchSnapshot();
  expect(await cache.get('test')).toMatchSnapshot();
  expect(await cache.get('test2')).toMatchSnapshot();
});

test('evict', async () => {
  expect(await cache.putIfAbsent('test', 'testnew')).toMatchSnapshot();
  expect(await cache.putIfAbsent('test2', 'test2')).toMatchSnapshot();
  expect(await cache.putIfAbsent('test3', 'test3')).toMatchSnapshot();
  expect(await cache.get('test3')).toMatchSnapshot();
  await cache.evict('test3');
  expect(await cache.get('test3')).toMatchSnapshot();
});

test('clear', async () => {
  expect(await cache.putIfAbsent('test', 'testnew')).toMatchSnapshot();
  expect(await cache.putIfAbsent('test2', 'test2')).toMatchSnapshot();
  expect(await cache.putIfAbsent('test3', 'test3')).toMatchSnapshot();
  expect(await cache.get('test')).toMatchSnapshot();
  expect(await cache.get('test2')).toMatchSnapshot();
  expect(await cache.get('test3')).toMatchSnapshot();
  await cache.clear();
  expect(await cache.get('test')).toMatchSnapshot();
  expect(await cache.get('test2')).toMatchSnapshot();
  expect(await cache.get('test3')).toMatchSnapshot();
});

function sleep(sec: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, sec * 1000));
}

test('put with duration', async () => {
  await cache2.put('test', 'test');
  expect(await cache2.get('test')).toMatchSnapshot();
  await sleep(3);
  expect(await cache2.get('test')).toMatchSnapshot();
});

test('get & put with timeout', async () => {
  expect(await cache3.get('test')).toMatchSnapshot();
  await cache3.put('test', 'test');
  expect(await cache3.get('test')).toMatchSnapshot();
  expect(await cache3.get('test')).toMatchSnapshot();

  const client = new Redis();
  client.debug('sleep', 2);
  await sleep(0.5);
  expect(await cache3.get('test')).toMatchSnapshot();
  await sleep(1.7);
  expect(await cache3.get('test')).toMatchSnapshot();

  await cache3.put('test', undefined);
  expect(await cache3.get('test')).toMatchSnapshot();
  await cache3.put('test2', null);
  expect(await cache3.get('test2')).toMatchSnapshot();

  client.disconnect();
});
