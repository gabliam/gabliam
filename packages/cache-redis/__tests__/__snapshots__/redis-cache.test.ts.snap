// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cache 1`] = `"test"`;

exports[`cache get & put 1`] = `undefined`;

exports[`cache get & put 2`] = `"test"`;

exports[`cache get & put 3`] = `"test"`;

exports[`cache get & put 4`] = `undefined`;

exports[`cache get & put 5`] = `null`;

exports[`cache putIfAbsent 1`] = `"test"`;

exports[`cache putIfAbsent 2`] = `null`;

exports[`cache putIfAbsent 3`] = `null`;

exports[`cache putIfAbsent 4`] = `"test"`;

exports[`cache putIfAbsent 5`] = `"test2"`;

exports[`clear 1`] = `null`;

exports[`clear 2`] = `null`;

exports[`clear 3`] = `null`;

exports[`clear 4`] = `"testnew"`;

exports[`clear 5`] = `"test2"`;

exports[`clear 6`] = `"test3"`;

exports[`clear 7`] = `undefined`;

exports[`clear 8`] = `undefined`;

exports[`clear 9`] = `undefined`;

exports[`evict 1`] = `null`;

exports[`evict 2`] = `null`;

exports[`evict 3`] = `null`;

exports[`evict 4`] = `"test3"`;

exports[`evict 5`] = `undefined`;

exports[`get & put with timeout 1`] = `undefined`;

exports[`get & put with timeout 2`] = `"test"`;

exports[`get & put with timeout 3`] = `"test"`;

exports[`get & put with timeout 4`] = `undefined`;

exports[`get & put with timeout 5`] = `"test"`;

exports[`get & put with timeout 6`] = `undefined`;

exports[`get & put with timeout 7`] = `null`;

exports[`put with duration 1`] = `"test"`;

exports[`put with duration 2`] = `undefined`;
